### ISSUE TITLE
p2p: forever connected to lazy peer with no progress

### ISSUE THREAD

**Issue by czarcas7ic** [cometbft/cometbft#3316]
As I was implementing the same geo peering feature, I started by peering to a single outbound peer to test that my feature was working as intended. With same geo peering, the pool of peers you fetch is much smaller. There was one peer that, any time I would connect to it, I would be stuck in the ensurePeers 30s check indefinitely. It seems there is no check to determine if a peer is "lazy", it only checks how many peers we are connected to. Just as a proof of concept I added the following logic branch to ensurePeers:

https://github.com/osmosis-labs/cometbft/blob/2c8bf088f89f6374a8eae3124060eabb4536ba60/p2p/pex/pex_reactor.go#L464-L469

When I connected to the lazy peer, we eventually disconnected, connected to a new peer, and began syncing. Obviously part of the solution here is "don't connect to only one peer", but we shouldn't be wasting peer slots on peers that do absolutely nothing in the first place.

Would like to brainstorm here on what the proper solution is. In my mind, what I did is close to what we want, but I think this Idle check should only be done on blocksync right? As in, during block sync, we really expect all the peers we are connected to to be sending us data. However, (this could be a wrong assumption), when we are caught up to the head, it would be acceptable for a peer to send us nothing for 30s, yet we would still want to be connected to them.

Please let me know your thoughts and I will be happy to upstream the desired solution.

As a side note, this does happen on the non geo peering branch, its just less likely to connect to a dead peer given the pool is every country rather than just the country the node is in.

**Comment by cason** [comment]
Interesting, but needs additional triage to understand.

The connection with a peer `MConnection` exchanges ping/pong messages with the connected peer. If responses are not received within the configured interval, the connection is closed:

https://github.com/cometbft/cometbft/blob/8e046e2c273c0feb9f3a0586c490ebefadfd3e70/p2p/conn/connection.go#L464-L467

The default Pong timeout is 45s.

**Comment by melekes** [comment]
> Would like to brainstorm here on what the proper solution is. In my mind, what I did is close to what we want, but this Idle check should only be done on blocksync right?

Well, not really. We want to discard lazy or outright malicious peers in all parts of the stack (block sync, state sync, consensus, mempool, etc.). There was always an idea of a peer scoring system, but we never implemented it. If I'm right, we only promote good peers (sending us block parts and other messages) and disconnect from "bad" peers, who send us invalid block parts and other stuff.

**Comment by melekes** [comment]
I'm not sure yet where the best place for this check is, but it's probably somewhere in `Switch`. @cason @sergio-mena, what do you think?

**Comment by cason** [comment]
So, again, we have a transport layer level 45 seconds timeout for idle connections. This is probably longer, because this timer starts when we send a ping, and we don't send pings all the time.

A second way was to conceive at p2p level (`Switch` probably) some mechanism to identify peers that are too slow, or lazy like you said, or even disconnected and we are not still aware of that (because the TCP connection is still there, and the pong timeout has not expired). I don't like this approach, to be honest.

The third way is to enable this "are-you-alive?" form of check in some reactors. Understandably, since a node typically starts from state or block sync, they are good candidates. But what happens, for instance, when we have a new network to which nodes are gradually joining? And we have a start time on the genesis file. Are we disconnecting just because they don't have anything useful to send?

**Comment by czarcas7ic** [comment]
As part of the scoring system, I notice that we store a peers avg bytes sent to you, but not how much you have requested from the peer. In other words, we don't want to penalize a peer for having low avg bytes if we are not requesting much from them, but should get rid of them quickly if they have lower avg bytes when we are requesting, which from an initial check I don't think we currently do (obviously not the full solution, but gets us closer to this "peer ranking" that I think will solve a lot of problems)

**Comment by czarcas7ic** [comment]
Has there been any design docs written on how a proper peer scoring system would be in comet? If that was broken down into clear issues I would be happy to take some of those on!

**Comment by cason** [comment]
> Has there been any design docs written on how a proper peer scoring system would be in comet? If that was broken down into clear issues I would be happy to take some of those on!

There were in the p2p refactoring work performed several years ago, that was at the end reverted. But still there the scoring system was pretty rough: https://github.com/tendermint/tendermint/issues/5670 and https://github.com/tendermint/tendermint/pull/6277.

**Comment by czarcas7ic** [comment]
@cason Is it possible the connect node is sending / receiving pings / pongs, but not actually sending any block data? I ask because, I had a node in this state for over an hour, connected to the same peer, and the amount of data sent was 0.

**Comment by cason** [comment]
Yes, it is possible. The ping/pong are I-am-alive messages, although also sent when we are sending data (usually, we only rely on these messages when the channel is idle, not like this in our implementation, unfortunately).

I think it is possible, as ping/pong messages are not counted as data exchanged. Probably, only full messages (all packets received) are counted.
